hbase-server/src/main/java/org/apache/hadoop/hbase/procedure2/store/region/RegionProcedureStore.java

```diff
@@ -0,0 +1,391 @@
+      }
+      region.mutateRowsWithLocks(mutations, rowsToLock, NO_NONCE, NO_NONCE);
+    } catch (IOException e) {
+      LOG.error(HBaseMarkers.FATAL, "Failed to insert proc {}, sub procs {}", proc,
+        Arrays.toString(subProcs), e);
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public void insert(Procedure<?>[] procs) {
+    List<Mutation> mutations = new ArrayList<>(procs.length);
+    List<byte[]> rowsToLock = new ArrayList<>(procs.length);
+    try {
+      for (Procedure<?> proc : procs) {
+        serializePut(proc, mutations, rowsToLock);
+      }
+      region.mutateRowsWithLocks(mutations, rowsToLock, NO_NONCE, NO_NONCE);
+    } catch (IOException e) {
+      LOG.error(HBaseMarkers.FATAL, "Failed to insert procs {}", Arrays.toString(procs), e);
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public void update(Procedure<?> proc) {
+    try {
+      ProcedureProtos.Procedure proto = ProcedureUtil.convertToProtoProcedure(proc);
+      region.put(new Put(Bytes.toBytes(proc.getProcId())).addColumn(FAMILY, PROC_QUALIFIER,
+        proto.toByteArray()));
+    } catch (IOException e) {
+      LOG.error(HBaseMarkers.FATAL, "Failed to update proc {}", proc, e);
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public void delete(long procId) {
+    try {
+      region.delete(new Delete(Bytes.toBytes(procId)));
+    } catch (IOException e) {
+      LOG.error(HBaseMarkers.FATAL, "Failed to delete proc id={}", procId, e);
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public void delete(Procedure<?> parentProc, long[] subProcIds) {
+    List<Mutation> mutations = new ArrayList<>(subProcIds.length + 1);
+    List<byte[]> rowsToLock = new ArrayList<>(subProcIds.length + 1);
+    try {
+      serializePut(parentProc, mutations, rowsToLock);
+      for (long subProcId : subProcIds) {
+        serializeDelete(subProcId, mutations, rowsToLock);
+      }
+      region.mutateRowsWithLocks(mutations, rowsToLock, NO_NONCE, NO_NONCE);
+    } catch (IOException e) {
+      LOG.error(HBaseMarkers.FATAL, "Failed to delete parent proc {}, sub proc ids={}", parentProc,
+        Arrays.toString(subProcIds), e);
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public void delete(long[] procIds, int offset, int count) {
+    if (count == 0) {
+      return;
+    }
+    if (count == 1) {
+      delete(procIds[offset]);
+      return;
+    }
+    List<Mutation> mutations = new ArrayList<>(count);
+    List<byte[]> rowsToLock = new ArrayList<>(count);
+    for (int i = 0; i < count; i++) {
+      serializeDelete(procIds[offset + i], mutations, rowsToLock);
+    }
+    try {
+      region.mutateRowsWithLocks(mutations, rowsToLock, NO_NONCE, NO_NONCE);
+    } catch (IOException e) {
+      LOG.error(HBaseMarkers.FATAL, "Failed to delete proc ids={}", Arrays.toString(procIds), e);
+      throw new UncheckedIOException(e);
+    }
+  }
+}
```

Comment:
Is that all it takes? That'd be great.

Reply:
Yes, much less code than before, the only overhead is that we need to implement the flush and compact schedule by ourselves.

Reply:
Sweet